[Serialization] Load non-public transitive dependencies on `@testable` imports

xymus opened the pull request:
A `@testable` import allows a client to call internal decls which may refer to non-public dependencies. To support such a use case, load non-public transitive dependencies of a module when it's imported `@testable` from the main module.

We apply this behavior only to `@testable` imports from the main module being built, not transitively trough `@testable` imports of a library. It should support the vast majority of `@testable` uses and provide a reasonable behavior without surprises.

This replaces the previous behavior where we loaded those dependencies for any modules built for testing. This would load more modules for any debug build, opening the door to a different behavior between debug and release builds. In contrast, applying this logic to `@testable` clients will only change the behavior of test targets.

rdar://107329303

---

The last commit, `[Serialization] Load non-public transitive dependencies on @testable imports`, is the most relevant to review this PR. The first 2 commits refactor existing code to allow loading dependencies in two phases.

xymus committed:
[Serialization|NFC] Extract logic loading dependencies out of associateWithFileContext

xymus committed:
[Serialization|NFC] Split diagnoseSerializedASTLoadFailure in two

The new diagnoseSerializedASTLoadFailureTransitive diagnose problems for
transitive dependencies only: missing dependency, missing underlying
module, or circular dependency.

xymus committed:
[Serialization] Load non-public transitive dependencies on @testable imports

A @testable import allows a client to call internal decls which may
refer to non-public dependencies. To support such a use case, load
non-public transitive dependencies of a module when it's imported
@testable from the main module.

This replaces the previous behavior where we loaded those dependencies
for any modules built for testing. This was risky as we would load more
module for any debug build, opening the door to a different behavior
between debug and release builds. In contrast, applying this logic to
@testable clients will only change the behavior of test targets.

rdar://107329303

xymus commented:
@swift-ci Please smoke test

elsh commented on test/Serialization/access-level-import-dependencies.swift:
```diff
@@ -95,4 +95,24 @@ import PrivateDep
+
+//--- TestableClientOfNonPublic.swift
+@testable import PackageDep // expected-error {{missing required module 'HiddenDep'}}
+@testable import InternalDep // expected-error {{missing required module 'HiddenDep'}}
```
Might be worth adding a test with `@_exported` import as well?

xymus replied:
I've added a test about it. From the commit message:

A scenario where module A `@testable` imports module B, which `@_exported` imports module C. In this case, non-public imports of B are loaded, but not those of C. This is a limitation of the current implementation and could be improved upon in the future.

bnbarham commented on lib/Serialization/ModuleFileSharedCore.h:
```diff
@@ -614,11 +614,17 @@ class ModuleFileSharedCore {
   /// loaded from the same package.
+  ///
+  /// If \p forTestable, get the desired loading behavior for a @testable
+  /// import. So it reports non-public dependencies as required for a testable
+  /// client can access internal details, which in turn can reference those
+  /// non-public dependencies.
```
I think the `so` is in the wrong position here, maybe:
```
<-So it> Reports ... as required for a testable client <+so it> can access
```

bnbarham commented:
Worth adding a WMO test where one file has a @testable import and the other doesn't?

xymus committed:
[Sema] Support -testable-import-module to load transitive non-public dependencies

When using the -testable-import-module argument to insert a testable
import, there's no ImportDecl on which to show the diagnostics when
loading transitive dependencies. Clean up the logic to still load
dependencies in such a case.

xymus committed:
[Serialization] Add 2 scenarios to the test for transitive testable dependencies

* A multi-file scenario where one file has a normal import and the other
  a testable import. In this case we still load the extra dependencies
  and show diagnostics only of the testable import.

* A scenario where module A @testable imports module B, which @_exported
  imports module C. In this case, non-public imports of B are loaded,
  but not those of C. This is a limitation of the current implementation
  and could be improved upon in the future.

xymus commented:
@swift-ci Please smoke test
